pkg/metadata/store/metadata_store.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package store
+
+import (
+	"sync"
+)
+
+// TODO(random-liu) Rename memoryMeta to metadata, and add checkpoint. When
+// checkpoint is enabled, it should cache data in memory and checkpoint metadata
+// into files during update. Metadata should serve from memory, but any modification
+// should be checkpointed, so that memory could be recovered after restart. It
+// should be possible to disable the checkpoint for testing.
+// Note that checkpoint update may fail, so the recovery logic should tolerant that.
+
+// An atomic file operator is needed.
+
+// memoryMeta is the in-memory implementation of Metadata.
+type memoryMeta struct {
+	sync.RWMutex
+	data []byte
+}
+
+var _ Metadata = &memoryMeta{}
+
+// newMemoryMeta creates a new memoryMetadta.
+func newMemoryMeta(data []byte) (*memoryMeta, error) {
+	return &memoryMeta{data: data}, nil
+	// TODO(random-liu): Create the data on disk atomically.
+}
+
+// Get a snapshot of the metadata.
+func (m *memoryMeta) Get() ([]byte, error) {
+	m.RLock()
+	defer m.RUnlock()
+	return m.data, nil
+}
+
+// Update the value.
+func (m *memoryMeta) Update(u UpdateFunc) error {
+	m.Lock()
+	defer m.Unlock()
+	newData, err := u(m.data)
+	if err != nil {
+		return err
+	}
+	// Replace with newData, user holding the old data will not
+	// be affected.
+	m.data = newData
+	return nil
+	// TODO(random-liu) Compare data to decide whether to update disk.
+	// TODO(random-liu) Hold read lock and **update** existing data on
+	// disk atomically.
+	// Only holding read lock makes sure that:
+	// 1) The in-memory cache is available for reading.
+	// 2) Even though, there is another on-going update, the read lock
+	// makes sure that they must be updating the same version of data.
+}
+
+// delete deletes the data on disk atomically.
+func (m *memoryMeta) delete() error {
+	// TODO(random-liu): Hold write lock, rename the data on the disk.
```

Comment:
Option 1: If we set `m.data=nil`, other operation (e.g. ListPodSandbox, PodSandboxStatus) which has already got the metadata reference will fail, and return error.

Option 2: If we don't set `m.data=nil`, other operation which has got the reference will success:
1) If that operation is read only, it will return stale data.
2) If that operation updates the metadata, it will only change the in-memory state.
3) Any future operations won't be able to get the metadata because it is removed from the list already.

Either option 1 or 2 is acceptable to me, because this is unavoidable race unless we lock the whole store for each single metadata operation, which is not what we want. :)

I prefer 2.


We don't necessarily need to set `m.data=nil`.

In cri-containerd, no long running process should hold a specific metadata reference,